libs/wingc/src/closure_transform.rs

```diff
@@ -0,0 +1,306 @@
+use std::cell::RefCell;
+
+use indexmap::IndexMap;
+
+use crate::{
+	ast::{
+		ArgList, Class, ClassField, Expr, ExprKind, FunctionDefinition, FunctionParameter, FunctionSignature, Initializer,
+		Phase, Reference, Scope, Stmt, StmtKind, Symbol, TypeAnnotation, TypeAnnotationKind, UserDefinedType,
+	},
+	diagnostic::WingSpan,
+	fold::{self, Fold},
+};
+
+const PARENT_THIS_NAME: &str = "__parent_this";
+
+/// Transforms inflight closures defined in preflight scopes into preflight classes.
+///
+/// This is done by wrapping the closure's code in a preflight class with a single method,
+/// and replacing the closure with a reference to an instance of that class.
+/// The class is given a unique name to avoid collisions.
+///
+/// For example, the following code:
+/// ```wing
+/// let b = new cloud.Bucket();
+/// let f = inflight (message: str) => {
+///   b.put("file.txt", message);
+/// };
+/// ```
+/// is transformed into:
+/// ```wing
+/// let b = new cloud.Bucket();
+/// class $Inflight1 {
+///   init() {}
+///   inflight handle(message: str) {
+///     b.put("file.txt", message);
+///   }
+/// }
+/// let f = new $Inflight1();
+pub struct ClosureTransformer {
+	// Whether the transformer is inside a preflight or inflight scope.
+	// Only inflight closures defined in preflight scopes need to be transformed.
+	curr_phase: Phase,
+	// Whether the transformer is inside a scope where "this" is valid.
+	inside_scope_with_this: bool,
+	// Helper state for generating unique class names
+	inflight_counter: usize,
+	// Stores the list of class definitions that need to be added to the nearest scope
+	class_statements: Vec<Stmt>,
+	// Track the statement index of the nearest statement we're inside so that
+	// newly-inserted classes will have access to the correct state
+	nearest_stmt_idx: usize,
+}
+
+impl ClosureTransformer {
+	pub fn new() -> Self {
+		Self {
+			curr_phase: Phase::Preflight,
+			inside_scope_with_this: false,
+			inflight_counter: 0,
+			class_statements: vec![],
+			nearest_stmt_idx: 0,
+		}
+	}
+}
+
+impl Fold for ClosureTransformer {
+	fn fold_scope(&mut self, node: Scope) -> Scope {
+		let mut statements = vec![];
+
+		// If we are inside a scope with "this", add define `let __parent_this = this` which can be
+		// used by the newly-created preflight classes
+		if self.inside_scope_with_this {
+			let parent_this_name = Symbol::global(PARENT_THIS_NAME); // TODO: can we use a span?
```

Comment:
Not that it solves the span thing but I recently added added a `From` trait to `Symbol` so this should be equivalent:
```suggestion
			let parent_this_name = PARENT_THIS_NAME.into(); // TODO: can we use a span?
```